corehq/apps/es/cases.py

```diff
@@ -53,9 +56,37 @@ class ElasticCase(ElasticDocumentAdapter):
     def mapping(self):
         return get_adapter_mapping(self)
 
-    @classmethod
-    def from_python(cls, doc):
-        return from_dict_with_possible_id(doc)
+    def from_python(self, case):
+        from corehq.form_processor.models.cases import CommCareCase
+        if isinstance(case, dict):
+            case_dict = copy(case)
+        elif isinstance(case, CommCareCase):
+            case_dict = case.to_json()
+        else:
+            raise TypeError(f"Unknown type {type(case)}")
+        return self._from_dict(case_dict)
+
+    def _from_dict(self, case):
+        """
+        Takes in case dict and applies required transformation to make it suitable for ES.
+        The function is replica of ``transform_case_for_elasticsearch``
+        In future all references to  ``transform_case_for_elasticsearch`` will be replaced by `from_python`
+
+        :param case: an instance of ``dict`` which is ``CommcareCase.to_json()``
+        """
+        from corehq.pillows.utils import get_user_type
+
+        if not case.get("owner_id"):
+            if case.get("user_id"):
+                case["owner_id"] = case["user_id"]
+
+        case['owner_type'] = get_user_type(case.get("owner_id", None))
+        case['inserted_at'] = datetime.utcnow().isoformat()
```

Comment:
pity this isn't called `@indexed_on` :(

Reply:
I know you just copied this but it should probably use `json_format_datetime`

Reply:
> pity this isn't called @indexed_on :(

Do you think it would be a good idea to add `@indexed_on` as a separate key and during reindex update `inserted_at` with `@indexed_on`?

> I know you just copied this but it should probably use json_format_datetime

Does `json_format_datetime` is timezone aware and add appends `Z` to denote UTC, wouldn't it result in having different `datetime` string formats for the same key? Should it be considered a problem or do you think its fine to update the formats? 

@snopoke

Reply:
> > pity this isn't called @indexed_on :(
> 
> Do you think it would be a good idea to add `@indexed_on` as a separate key and during reindex update `inserted_at` with `@indexed_on`?

It would be nice for consistency but we'd also then need to make changes to APIs that reference the `inserted_at` so it would be a larger task and I'm not sure it's worth it.


> > I know you just copied this but it should probably use json_format_datetime
> 
> Does `json_format_datetime` is timezone aware and add appends `Z` to denote UTC, wouldn't it result in having different `datetime` string formats for the same key? Should it be considered a problem or do you think its fine to update the formats?

Technically I think it's an improvement since it's now clear that the date is in UTC. Under the hood ES converts dates to UTC for storage and any date without a TZ is assumed to be UTC so I don't think it would cause any issue.

Reply:
https://github.com/dimagi/commcare-hq/pull/32660/commits/c660f9b6bd7f7f54eb22e8ad1e68bdec219ccaec